applications/ConstitutiveLawsApplication/custom_constitutive_thermal/thermal_elastic_isotropic_3d.cpp

```diff
@@ -0,0 +1,151 @@
+// KRATOS ___                _   _ _         _   _             __                       _
+//       / __\___  _ __  ___| |_(_) |_ _   _| |_(_)_   _____  / /  __ ___      _____   /_\  _ __  _ __
+//      / /  / _ \| '_ \/ __| __| | __| | | | __| \ \ / / _ \/ /  / _` \ \ /\ / / __| //_\\| '_ \| '_  |
+//     / /__| (_) | | | \__ \ |_| | |_| |_| | |_| |\ V /  __/ /__| (_| |\ V  V /\__ \/  _  \ |_) | |_) |
+//     \____/\___/|_| |_|___/\__|_|\__|\__,_|\__|_| \_/ \___\____/\__,_| \_/\_/ |___/\_/ \_/ .__/| .__/
+//                                                                                         |_|   |_|
+//
+//  License:         BSD License
+//                   license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Alejandro Cornejo
+//
+
+// System includes
+#include <iostream>
+
+// External includes
+
+// Project includes
+#include "custom_constitutive_thermal/thermal_elastic_isotropic_3d.h"
+
+#include "includes/checks.h"
+#include "structural_mechanics_application_variables.h"
+
+namespace Kratos
+{
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void ThermalElasticIsotropic3D::CalculateMaterialResponsePK2(ConstitutiveLaw::Parameters& rValues)
+{
+    KRATOS_TRY
+    const auto& r_process_info = rValues.GetProcessInfo();
+    if (r_process_info[STEP] == 1) // We storage the ref temperature as the initial one
+        mReferenceTemperature = AdvancedConstitutiveLawUtilities<6>::CalculateInGaussPoint(TEMPERATURE, rValues);
```

Comment:
this should be done in `Initialize` (and check for restart)

Reply:
I agree with @philbucher

Reply:
Yeah I thought about doing that. It is a solution depending on your needs. In this case I adopted that the ref Temp was the one at the very beginning of the calculation (can be different FE wise). In this case, **it is required that the Thermal solver at leas has computed the Temp field once**, and this is after the `Initialize `method I think.

Reply:
The reference Temp could be also a material property BTW

Reply:
Can be a property... I am thinking that you maybe can define the temp information as a pointer to a struct, and check if the pointer is null (so if it is not possible to initialize in the Initialize can be checked later)... I don't know, there are many options. Can be just simply be set with the "SetValue" method, and be set outside, from the reading processes or some other point of the solver. There are many options

Reply:
Opinions @philbucher ? (in any case  the SetValue should be implemented)

Reply:
I am not sure if we need such a complicated solution for the ref temp
I think having `REFERENCE_TEMPERATURE` would be sufficient
```c++
// in Initialize
if properties.Has(REFERENCE_TEMPERATURE)
    mReferenceTemperature = REFERENCE_TEMPERATURE;
else
    mReferenceTemperature = TEMPERATURE;
```

or is that too simple?

Reply:
yes, I'm going to implement the option with the SetValue

Reply:
@philbucher This will be the simplest option, but if we want a ref temp element-wise, we should add as many material properties as FE! :D

Reply:
ah
in this case I suggest:
```c++
// in Initialize
if this->Has(REFERENCE_TEMPERATURE)
    mReferenceTemperature = REFERENCE_TEMPERATURE;
elif properties.Has(REFERENCE_TEMPERATURE)
    mReferenceTemperature = REFERENCE_TEMPERATURE;
else
    mReferenceTemperature = TEMPERATURE;
```

Reply:
exactly, this is what I'm going to implement. The last option is the problematic in this case because the TEMPERATURE is not computed yet (Initialize state). I'll try to circumvent this problem

Reply:
We can add a new flag ... (opening Pandora's box) 

TEMPERATURE_INITIALIZED for example

Reply:
> We can add a new flag ... (opening Pandora's box)
> 
> TEMPERATURE_INITIALIZED for example

I see my proposal is not welcome. We can just simply modify the reading material process, so this is set on that reading fase... In any case ould be troublesome